lib/ex_unit/lib/ex_unit/runner.ex

```diff
@@ -299,14 +351,78 @@ defmodule ExUnit.Runner do
                 type: Atom.to_string(test.tags.test_type)
               )
 
-            %{test | state: failed(:error, exception, stacktrace)}
+            test = %{test | state: failed(:error, exception, stacktrace)}
+            process_failure(test, options)
 
           nil ->
-            receive_test_reply(test, test_pid, test_ref, timeout)
+            receive_test_reply(test, options)
         end
     end
   end
 
+  defp process_failure(
+         %ExUnit.TestModule{state: {tag, _}, tests: tests} = test_module,
+         %{pid: module_pid, parent_pid: parent_pid, config: config} = _options
+       )
+       when tag in [:failed, :invalid] do
+    failure_counter = increment_failure_counter(config.manager, test_module, length(tests))
+
+    if max_failures_reached?(failure_counter, config.max_failures) do
+      terminate_all_running_processes(config.manager, parent_pid, module_pid)
+      {:error, {:max_failures_reached, test_module}}
+    else
+      {:ok, test_module}
+    end
+  end
+
+  defp process_failure(%ExUnit.TestModule{} = test_module, _options) do
+    {:ok, test_module}
+  end
+
+  defp process_failure(
+         %ExUnit.Test{state: {tag, _}} = test,
+         %{pid: test_pid, parent_pid: parent_pid, config: config} = _options
+       )
+       when tag in [:failed, :invalid] do
+    failure_counter = increment_failure_counter(config.manager, test)
+
+    if max_failures_reached?(failure_counter, config.max_failures) do
+      terminate_all_running_processes(config.manager, parent_pid, test_pid)
+      {:error, {:max_failures_reached, test}}
+    else
+      {:ok, test}
+    end
+  end
+
+  defp process_failure(%ExUnit.Test{} = test, _options) do
+    {:ok, test}
+  end
+
+  # Stops all running tests in the suite
+  defp terminate_all_running_processes(manager, parent_pid, current_pid) do
+    for registered_pid <- ExUnit.OnExitHandler.get_registered_pids(manager),
+        registered_pid not in [parent_pid, current_pid],
+        Process.alive?(registered_pid) do
+      Process.exit(registered_pid, :max_failures_reached)
+    end
+  end
+
+  defp get_failure_counter(manager),
+    do: ExUnit.OnExitHandler.get_failure_counter(manager)
+
+  defp increment_failure_counter(manager, test, increment \\ 1)
```

Comment:
[`ExUnit.OnExitHandler.increment_failure_counter/2`](https://github.com/elixir-lang/elixir/pull/8174/files#diff-258d820db9fe3b2672410c8255211706R88) also validates the `increment` argument, so we might not need to also check for it here.

This internal function is only called from [here](https://github.com/elixir-lang/elixir/pull/8174/files#diff-1e5179d66aaabef6fbc8efa93ae33493R368) and [here](https://github.com/elixir-lang/elixir/pull/8174/files#diff-1e5179d66aaabef6fbc8efa93ae33493R387). Keeping that in mind,  I think the last clause  `defp increment_failure_counter(manager, _, _increment)` won't be executed; and the `tag` on `ExUnit.TestModule` won't never be `:not_executed` 🙂

I think everything could be simplified to (the `tag` is checked on the parent calls):

```elixir
defp increment_failure_counter(manager, increment \\ 1) do
  ExUnit.OnExitHandler.increment_failure_counter(manager, increment)
end
```